backend/app/notify/email.go

```diff
@@ -0,0 +1,411 @@
+package notify
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net"
+	"net/smtp"
+	"sync"
+	"text/template"
+	"time"
+
+	log "github.com/go-pkgz/lgr"
+	"github.com/go-pkgz/repeater"
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+)
+
+// EmailParams contain settings for email notifications
+type EmailParams struct {
+	From                 string        // From email field
+	MsgTemplate          string        // request message template
+	VerificationSubject  string        // verification message subject
+	VerificationTemplate string        // verification message template
+	BufferSize           int           // email send buffer size
+	FlushDuration        time.Duration // maximum time after which email will me sent, 30s by default
+}
+
+// Email implements notify.Destination for email
+type Email struct {
+	EmailParams
+	SmtpParams
+
+	smtp       smtpClientWithMaker
+	msgTmpl    *template.Template // parsed request message template
+	verifyTmpl *template.Template // parsed verification message template
+	submit     chan emailMessage  // unbuffered channel for email sending
+	once       sync.Once
+}
+
+// SmtpParams contain settings for smtp server connection
+type SmtpParams struct {
+	Host     string        // SMTP host
+	Port     int           // SMTP port
+	TLS      bool          // TLS auth
+	Username string        // user name
+	Password string        // password
+	TimeOut  time.Duration // TCP connection timeout
+}
+
+// default email client implementation
+type emailClient struct{ smtpClientWithMaker }
+
+type smtpClientWithMaker interface {
+	smtpClient
+	smtpMaker
+}
+
+// smtpClient interface defines subset of net/smtp used by email client
+type smtpClient interface {
+	Mail(string) error
+	Auth(smtp.Auth) error
+	Rcpt(string) error
+	Data() (io.WriteCloser, error)
+	Quit() error
+	Close() error
+}
+
+// smtpMaker interface defines function for creating new smtpClients
+type smtpMaker interface {
```

Comment:
minor: a single-method interface better to be in sync with named, i.e. for blahMaker the mathod better be named Make. If you want to keep it create maybe just call it smtpCreator

Reply:
Excellent point, now `smtpClientWithCreator` contains  `smtpClientCreator`.